File: Recursion-Backtracking/05-Combination-Sum-II/combination-sum-ii.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */

function findPermutations(index, n, curr_sum, target, curr, result, m, candidates){
    if(curr_sum === target){
        const resultArr = [...curr];
        const sortedArr = resultArr.sort((a,b) => a - b);
        const key = JSON.stringify(sortedArr);
        if(!m.hasOwnProperty(key)){
            m[key] = true;
            result.push(sortedArr);
        }
        return;
    }

    if(index === n || curr_sum > target) return;

    curr.push(candidates[index]);
    findPermutations(index + 1, n, curr_sum + candidates[index], target, curr, result, m, candidates);
    curr.pop();
    let element = candidates[index];
    while(index < n && element === candidates[index]){
        index++;
    }
    findPermutations(index, n, curr_sum, target, curr, result, m, candidates);
}

var combinationSum2 = function(candidates, target) {
    const result = [];
    const curr = [];
    const m = {};

    findPermutations(0, candidates.length, 0, target, curr, result, m, candidates);

    return result;
};

File: Recursion-Backtracking/05-Combination-Sum-II/combination-sum-ii.cpp
class Solution {
public:

    void findCombinations(int index, int n, int curr_sum, int target, vector<int>& candidates, set<vector<int>>& s, vector<int>& curr){
        if(curr_sum == target){
            vector<int> temp(curr.begin(), curr.end());
            sort(temp.begin(), temp.end());
            s.insert(temp);
            return;
        }

        if(index == n) return;
        if(curr_sum > target) return;

        curr.push_back(candidates[index]);
        findCombinations(index + 1, n, curr_sum + candidates[index], target, candidates, s, curr);
        curr.pop_back();
        int element = candidates[index];
        while(index < n && candidates[index] == element){
            index++;
        }
        findCombinations(index, n, curr_sum, target, candidates, s, curr);
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        set<vector<int>> s;
        vector<int> curr;

        findCombinations(0, candidates.size(), 0, target, candidates, s, curr);

        vector<vector<int>> result(s.begin(), s.end());
        return result;
    }
};